### PR TITLE
upgrade(installer): Add verify methods for ld.so.preload & docker/daemon.json

### PR DESCRIPTION
**Pull request by BaptisteFoy**
### What does this PR do?
Adds verify methods for the two files that are edited when installing the injector:
- /etc/ld.so.preload
- /etc/docker/daemon.json
  - Note: only works for docker > 23.0.0

For docker we also verify post-reload that the runtimes have been changed to what we expect (default-runtime == dd-shim)

Also adds corresponding E2Es

### Motivation
Making sure nothing breaks on the host

### Additional Notes
N/A

### Possible Drawbacks / Trade-offs
N/A

### Describe how to test/QA your changes
N/A

### CODE REVIEW

**Review by coignetp — APPROVED** [review]

File: pkg/fleet/installer/service/apm_inject.go [hunk 1]
```diff
@@ -216,6 +217,27 @@ func (a *apmInjectorInstaller) deleteLDPreloadConfigContent(ldSoPreload []byte)
...
+// 1. Parsing it
+// 2. Using the LD_PRELOAD env var, checking every single lib in the file to open a simple shell
+func (a *apmInjectorInstaller) verifyLDPreloadFile(path string) error {
+	file, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	libs := strings.Fields(string(file))
+	for _, lib := range libs {
+		cmd := exec.Command("sh", "-c", "/bin/echo 1")
```
Comment (arbll) [inline pkg/fleet/installer/service/apm_inject.go:230]: can we remove the shell ? it adds a bunch of failure scenarios (simple example: their profile is broken)

File: pkg/fleet/installer/service/docker.go [hunk 2]
```diff
@@ -117,6 +127,67 @@ func (a *apmInjectorInstaller) deleteDockerConfigContent(previousContent []byte)
...
+	if len(dockerVersionSplit) < 3 {
+		log.Warnf("failed to parse docker version %s, skipping verification", versionBuffer.String())
+	}
+	majorDockerVersionInt, err := strconv.Atoi(dockerVersionSplit[0])
+	if err != nil {
+		log.Warnf("failed to parse docker version %s: %s, skipping verification", dockerVersionSplit[0], err)
+		return nil
+	}
+	if majorDockerVersionInt < 23 {
+		log.Warnf("docker version %d is not supported for verification, skipping", majorDockerVersionInt)
```
Comment (arbll) [inline pkg/fleet/installer/service/docker.go:153]: debug

File: pkg/fleet/installer/service/docker.go [hunk 3]
```diff
@@ -117,6 +127,67 @@ func (a *apmInjectorInstaller) deleteDockerConfigContent(previousContent []byte)
...
+		return nil
+	}
+
+	cmd = exec.Command("dockerd", "--validate", "--config-file", path)
+	buf := new(bytes.Buffer)
```
Comment (arbll) [inline pkg/fleet/installer/service/docker.go:158]: :nice:

File: pkg/fleet/installer/service/docker.go [hunk 4]
```diff
@@ -117,6 +127,67 @@ func (a *apmInjectorInstaller) deleteDockerConfigContent(previousContent []byte)
...
+// As the reload is eventually consistent we have to retry a few times
+func (a *apmInjectorInstaller) verifyDockerRuntime() (err error) {
+	for i := 0; i < 3; i++ {
+		if i > 0 {
+			time.Sleep(time.Second)
+		}
+		cmd := exec.Command("docker", "system", "info", "--format", "{{ .DefaultRuntime }}")
+		var outb bytes.Buffer
+		cmd.Stdout = &outb
+		err = cmd.Run()
```
Comment (arbll) [inline pkg/fleet/installer/service/docker.go:179]: Why are you not checking the error under here 🤔

File: pkg/fleet/installer/service/docker.go [hunk 5]
```diff
@@ -117,6 +127,67 @@ func (a *apmInjectorInstaller) deleteDockerConfigContent(previousContent []byte)
...
+// verifyDockerRuntime validates that docker runtime configuration contains
+// a path to the injector runtime. This must be done because we can't always
+// verify the configuration befor a reload depending on the docker version
+// As the reload is eventually consistent we have to retry a few times
+func (a *apmInjectorInstaller) verifyDockerRuntime() (err error) {
+	for i := 0; i < 3; i++ {
+		if i > 0 {
+			time.Sleep(time.Second)
+		}
+		cmd := exec.Command("docker", "system", "info", "--format", "{{ .DefaultRuntime }}")
```
[the remark this answers is not in the document]
Comment (BaptisteFoy) [reply pkg/fleet/installer/service/docker.go:176]: 17.03 at least

File: pkg/fleet/installer/service/docker.go (hunk 4 above)
Comment (BaptisteFoy) [reply pkg/fleet/installer/service/docker.go:179]: We're retrying it and only checking the last one. I will add a log.

**Review by arbll — APPROVED** [review]

**Pull request merged**